File: 2024/jump.c
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>

bool canJump(int* nums, int numsSize) {
    int max_reach = 0;

    if (numsSize == 1) {
        return true;
    }

    for (int i = 0; i < numsSize; i++)
    {
        int reachable = i + nums[i];

        if (max_reach >= i && nums[i] < 1) {
            if (max_reach >= numsSize - 1) {
                return true;
            }
            return false;
        }

        if (reachable > max_reach) {
            max_reach = reachable;
        }
    }

    if (max_reach >= numsSize - 1) {
        return true;
    }
    else {
        return false;
    }
}

int main() {
    int nums[] = {2, 0}; // Example array, modify as needed
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    if (canJump(nums, numsSize)) {
        printf("Can jump to the end.\n");
    } else {
        printf("Cannot jump to the end.\n");
    }

    return 0;
}

File: 2024/jump.py
from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        dp = [False] * len(nums)
        
        for i in range(len(nums)):
            if i == 0:
                dp[i] = True
            else:
                j = i - 1
                while j >= 0:
                    if dp[j] and nums[j] >= i - j:
                        dp[i] = True
                        break
                    j -= 1

        return dp[-1]                


s = Solution()
# print(s.canJump([2,3,1,1,4]))
print(s.canJump([3,2,1,0,4]))

